packages/neptune-extension/src/common/state/notebook/actions.ts

```diff
@@ -1,13 +1,116 @@
+import { ThunkAction, ThunkDispatch } from 'redux-thunk'
+import { AnyAction } from 'redux';
+import {CheckpointDTO, NotebookDTO, NotebookWithNoContentDTO} from 'generated/leaderboard-client/src/models';
+import { leaderboardClient } from "common/api/leaderboard-client";
 
-import { NotebookDTO } from 'generated/leaderboard-client/src/models';
+export const fetchNotebook = (notebookId: string): ThunkAction<Promise<NotebookDTO | void>, {}, {}, NotebookActions> => {
+  return (dispatch: ThunkDispatch<{}, {}, AnyAction>) => {
+    return leaderboardClient.api
+      .getNotebook({ id: notebookId })
+      .then((notebook) => {
+        dispatch(fetchNotebookSuccess(notebook));
+        return notebook;
+      })
+      .catch(() => {
+        dispatch(fetchNotebookFailed());
+        return;
+      })
+  };
+};
 
-export function setNotebook(notebook?: NotebookDTO) {
-  return {
-    type: 'SET_NOTEBOOK',
-    payload: {
-      notebook,
-    },
-  } as const;
+function fetchNotebookSuccess(notebook: NotebookDTO) {
+  return { type: 'NOTEBOOK_FETCH_SUCCESS', payload: {notebook} } as const;
 }
 
-export type NotebookActions = ReturnType<typeof setNotebook>
+function fetchNotebookFailed() {
+  return { type: 'NOTEBOOK_FETCH_FAILED'} as const;
+}
+
+interface CheckpointMetadata {
+  path: string
+  name: string
+  description: string
+}
+
+export const uploadNotebook = (projectId: string, checkpointMeta: CheckpointMetadata, content: any)
+  : ThunkAction<Promise<NotebookWithNoContentDTO | void>, {}, {}, NotebookActions> => {
+  return (dispatch: ThunkDispatch<{}, {}, AnyAction>) => {
+
+    dispatch(uploadNotebookRequest());
+
+    return leaderboardClient.api
+      .createNotebook({ projectIdentifier: projectId })
+      .then((notebook) => {
+        dispatch(uploadNotebookSuccess(notebook))
```

Comment:
This is premature according to my knowledge, because the process in only complete when *both* notebook and checkpoint gets uploaded successfuly.